Oidc v2

LBolzani opened the pull request:
Hello Peter, I've completed the OIDC implementation starting from your oidc branch.
Could you make a check?

I remain at your disposal for any further information.

Luca

peterbanda commented:
Hey Luca,

I am currently on travel but will take a look in a week or so. Meanwhile pls. make sure there are no superfluous / unrelated changes (e.g. reorganization of the imports in AbstractCacheAsyncCrudRepo.scala and CustomSqlFunctions.java) as well as changes in the code indent (e.g. User.scala). But most importantly it would help me a lot if you could specify shortly what you have tried to achieve and was missing in the previous version. Thanx!

LBolzani commented:
Good morning Peter,

I've fixed the code avoiding superfluous changes.
Essentially my updates refer to the identification process.
Having different IdPs we can't rely on _userId_ only, but the unique identifier is _oidcId_ (the "sub" in jwt token).
The _oidcId_ is a unique common fields between the IdPs.

I've changed a little bit the logic in _OidcAuthController_ in order to consider the users present in database and the new ones.

When you're coming back I would like to plan a meeting to ask some questions about these topics:

- Production/release deployment;
- Database migration;
- Testing.

I remain at your disposal for any further information.

Luca

peterbanda commented on project/plugins.sbt:
```diff
@@ -41,3 +41,7 @@ addSbtPlugin("com.typesafe.sbt" % "sbt-mocha" % "1.0.0")
+
+
+// util
+addSbtPlugin("au.com.onegeek" %% "sbt-dotenv" % "2.1.146")
```
Not related to the PR... revert

peterbanda commented on server/build.sbt:
```diff
@@ -15,6 +15,8 @@ val playVersion = "2.5.9"
 libraryDependencies ++= Seq(
   "com.typesafe.play" %% "play" % playVersion,
   "com.typesafe.play" %% "play-json" % playVersion,
+  "org.pac4j" % "play-pac4j" % "3.1.0",
```
PAC4J and PAC4J OIDC are client/web libraries so the lib imports should be moved back to `web/build.sbt`

I think you needed it because of the tests, in particular the container `GuicePlayTestApp` class which is indeed in the server project. Nevertheless for testing strictly _client_-related stuff one should introduce a similar "test" container in the web project as well (with all necessary add-ons).

peterbanda commented on server/src/main/scala/org/ada/server/models/User.scala:
```diff
@@ -6,20 +6,26 @@ import play.api.libs.json.{Json, _}
-  ldapDn: String, // TODO: rename to userName
+  _id: Option[BSONObjectID] = None,
+  userId: String,
+  oidcId: Option[UUID] = None,
```
OIDC ID should be typed as `Option[String]` (as it was before), since according to the specification "sub" in jwt token can be any String (e.g. shown [here](https://auth0.com/docs/tokens/json-web-tokens/json-web-token-claims)).

Note that the use of `UUID` is specific to the LCSB deployment (with underlying Keycloak). Furthermore, this type restriction does not _add_ anything since it is completely ignored in the business logic (it's basically just passed around).

peterbanda commented on server/src/main/scala/org/ada/server/services/GuicePlayApp.scala:
```diff
@@ -15,8 +16,13 @@ object GuicePlayTestApp {
         import scala.collection.JavaConversions.iterableAsScalaIterable
         config.getStringList("play.modules.enabled").fold(
-          List.empty[String])(l => iterableAsScalaIterable(l).toList)
+          List.empty[String])(l => iterableAsScalaIterable(l).filterNot(_ == "org.ada.web.security.PacSecurityModule").toList)
```
This looks hacky... remove the `filterNot` part and simply don't pass the `PacSecurityModule`

peterbanda commented on server/src/main/scala/org/ada/server/services/GuicePlayApp.scala:
```diff
@@ -15,8 +16,13 @@ object GuicePlayTestApp {
       } else {
```
As I mentioned in the comment regarding the import of PAC4J libraries in the server project, it is better to extend `GuicePlayTestApp` and create a new test container class in the `web` project with all the necessary web/client stuff (such as `PlayCacheSessionStore`).

peterbanda commented on server/src/main/scala/org/ada/server/services/UserManager.scala:
```diff
@@ -31,10 +33,16 @@ trait UserManager {
 
   def findByEmail(email: String): Future[Option[User]]
 
+  def findByOidcId(oidcOpt: Option[UUID]): Future[Option[User]]
```
The function `findByOidcId` is not needed... will be elaborated down in the review

peterbanda commented on server/src/main/scala/org/ada/server/services/UserManager.scala:
```diff
@@ -161,5 +169,13 @@ private class UserManagerImpl @Inject()(
+    * @param oidcId to be matched
+    * @return Option containing Account with matching oidcId; None otherwise
+    */
+  override def findByOidcId(oidcOpt: Option[UUID]): Future[Option[User]] =
```
Should be removed... `oidcId` is optional (can be None/undefined) and therefore the field shouldn't be used (as it stands right now) as the core id to identify users

peterbanda commented on server/src/test/scala/stats/EigenVectorsTest.scala:
```diff
@@ -23,7 +23,7 @@ class EigenVectorsTest extends AsyncFlatSpec with Matchers with ExtraMatchers {
   private val precision = 0.0001
 
-  private val injector = GuicePlayTestApp().injector
+  private val injector = GuicePlayTestApp(addPlayCache = false).injector
```
Once `GuicePlayTestApp` is _server-only_ the `addPlayCache` can be removed

peterbanda commented on server/src/test/scala/stats/MetricMDSTest.scala:
```diff
@@ -31,7 +31,7 @@ class MetricMDSTest extends AsyncFlatSpec with Matchers with ExtraMatchers {
   private val precision = 0.001
 
-  private val injector = GuicePlayTestApp().injector
+  private val injector = GuicePlayTestApp(addPlayCache = false).injector
```
Once `GuicePlayTestApp` is _server-only_ the `addPlayCache` can be removed

peterbanda commented on web/app/org/ada/web/controllers/AppController.scala:
```diff
@@ -58,10 +58,10 @@ class AppController @Inject() (
           val dataSpacesNum = metaInfos.map(dataSpaceService.countDataSpacesNumRecursively).sum
           val dataSetsNum = metaInfos.map(dataSpaceService.countDataSetsNumRecursively).sum
-          val userFirstName = user.ldapDn.split("\\.", -1).head.capitalize
+          val userFirstName = user.userId.split("\\.", -1).head.capitalize
```
Instead of `user.userId.split("\\.", -1).head.capitalize` we can use now `user.name` (first part)

peterbanda commented on web/app/org/ada/web/controllers/AuthController.scala:
```diff
@@ -136,15 +159,15 @@ class AuthController @Inject() (
   def loginBasic = Action.async { implicit request =>
     if(userManager.debugUsers.nonEmpty)
-      gotoLoginSucceeded(userManager.basicUser.ldapDn)
+      gotoLoginSucceeded(userManager.basicUser.oidcId.get.toString)
```
This fails because `adminUser` does not have `oidcId` defined (see `User(userId = "admin.user", name = "Dummy Admin User", email = "admin@mail", roles = Seq("admin"))`).

The more general reason is that the `oidcId` field (as it stands right now) cannot be used as the core id, simply because it's optional. There are for instance LDAP or local Ada deployments where no OIDC auth is used...

Change to `basicUser.userId`

peterbanda commented on web/app/org/ada/web/controllers/AuthController.scala:
```diff
@@ -136,15 +159,15 @@ class AuthController @Inject() (
   def loginAdmin = Action.async { implicit request =>
     if (userManager.debugUsers.nonEmpty)
-      gotoLoginSucceeded(userManager.adminUser.ldapDn)
+      gotoLoginSucceeded(userManager.adminUser.oidcId.get.toString)
```
Similar to above pls. change to `adminUser.userId`

peterbanda commented on web/app/org/ada/web/security/PacSecurityModule.scala:
```diff
@@ -0,0 +1,61 @@
+
+    // callback
+    val callbackController = new CallbackController()
+    callbackController.setDefaultUrl("/?defaulturlafterlogout")
```
Remove `callbackController.setDefaultUrl("/?defaulturlafterlogout")`

peterbanda commented on web/app/org/ada/web/security/PacSecurityModule.scala:
```diff
@@ -0,0 +1,61 @@
+    // logout
+    val logoutController = new LogoutController()
+    logoutController.setDefaultUrl(adaBaseUrl)
+    logoutController.setCentralLogout(true)
```
Because of the never-ending discussion regarding the local-vs-global logout pls. revert the `enableCentralLogout` config (as it was originally done)

peterbanda commented on web/app/org/ada/web/security/PacSecurityModule.scala:
```diff
@@ -0,0 +1,61 @@
+    oidcConfiguration.setSecret(configuration.getString("oidc.secret").get)
+    oidcConfiguration.setClientAuthenticationMethod(ClientAuthenticationMethod.CLIENT_SECRET_POST)
+    oidcConfiguration.setPreferredJwsAlgorithm(JWSAlgorithm.RS256)
+    oidcConfiguration.setDiscoveryURI(configuration.getString("oidc.discoveryURI").get)
```
Revert all the configuration options as they were before, regardless of the current LCSB-specific deployment (with Keycloak). Other facilities that aim to use the OIDC auth will likely have different configs so it's better to be as general as possible here.

The following config options should be reintroduced:
- `logoutURL, responseType, scope, clientAuthenticationMethod, preferredJwsAlgorithm` and `useNonce`

peterbanda commented on web/app/org/ada/web/security/PacSecurityModule.scala:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  @Provides
+  def provideConfig(oidcClient: OidcClient[OidcProfile]) : Config = {
```
As mentioned before, the use of OIDC is optional. The original block (as it was implemented before the OIDC v2 was started) with a safe creation of `oidcClient` based on the presence (or absence) of `oidc.baseUrl` (or as it's now with `oidc.discoveryURI`) should be restored. You don't want to create an OIDC client if there is no OIDC configuration present (i.e., LDAP or local auth is used).

peterbanda commented on web/app/org/ada/web/security/PacSecurityModule.scala:
```diff
@@ -0,0 +1,61 @@
+
+  @Provides
+  def provideConfig(oidcClient: OidcClient[OidcProfile]) : Config = {
+    val clients = new Clients(s"$adaBaseUrl/oidc/callback", oidcClient)
```
Instead of the hardcoded url `/oidc/callback` which can be potentially changed in `routes` we should use here the original construction:
```
      val suffix = org.pac4j.play.routes.CallbackController.callback().url
      val callbackUrl = baseUrl.replaceAll("/$", "") + suffix
```

peterbanda commented on web/app/org/ada/web/security/PacSecurityModule.scala:
```diff
@@ -0,0 +1,61 @@
+  def provideOidcClient: OidcClient[OidcProfile] = {
+    val oidcConfiguration = new OidcConfiguration()
+    oidcConfiguration.setClientId(configuration.getString("oidc.clientId").get)
+    oidcConfiguration.setSecret(configuration.getString("oidc.secret").get)
```
All the `Option.get` calls are bad and should be avoided. If these config settings are not specified by mistake or intentionally because no OIDC auth was needed  the `.get` calls will fail without a reasonable exception. 

Instead of that pls. use `.getOrElse(throw new AdaException(s"OIDC config key '${key}' not available."))` (as it was originally done).

peterbanda commented on web/app/views/navbar.scala.html:
```diff
@@ -125,7 +125,7 @@
                         </li>
                     }{
-                        <a href="@routes.AuthController.login">Login&#160;&#160;&#160;<span class="glyphicon glyphicon-log-in"></span></a>
+                        <a href="@routes.OidcAuthController.oidcLogin()">Login&#160;&#160;&#160;<span class="glyphicon glyphicon-log-in"></span></a>
```
This is a bug...pls. revert. The original `routes.AuthController.login` contains a switch either to the classic login or to the OIDC one based on the configuration.

peterbanda commented on web/conf/core/auth.conf:
```diff
@@ -16,4 +17,14 @@ ldap {
+  adaBaseUrl = ${?ADA_BASE_URL}
+
+  returnAttributeIdName = "sub"
 }
```
As mentioned above... restore (or document) all the configuration options regardless the fact that the default values are currently sufficient at the test deployment. These are:

```
  clientAuthenticationMethod = "client_secret_post"    // optional
  preferredJwsAlgorithm = "RS256"                      // optional
//  responseType = "code" // id_token token            // optional
//  scope = "openid email profile phone"               // optional
//  useNonce = "true"                                  // optional
//  enableCentralLogout = false                        // optional - true by default
```

peterbanda commented on web/test/scala/org/ada/server/controller/AuthControllerTest.scala:
```diff
@@ -0,0 +1,34 @@
+
+class AuthControllerTest extends PlaySpec {
+
+  implicit val app: Application = GuicePlayTestApp()
```
Use `GuicePlayWebTestApp` (must be created first) here

peterbanda commented on web/app/org/ada/web/controllers/OidcAuthController.scala:
```diff
@@ -0,0 +1,119 @@
+import javax.inject.Inject
+import scala.concurrent.Future
+
+class OidcAuthController @Inject() (
```
This needs to be refactored...see my last comment for a possible solution

peterbanda commented on web/app/org/ada/web/controllers/UserController.scala:
```diff
@@ -187,7 +189,7 @@ class UserController @Inject() (
     } yield {
       val idAndNames = users.toSeq.map( user =>
-        Json.obj("_id" -> user._id, "name" -> user.ldapDn)
+        Json.obj("_id" -> user._id, "name" -> user.userId)
```
Once we switch the semantics of `userId` such that it can hold any id of any (current or future) auth provider (not just LDAP) we should not assume it's human readable and instead we should switch to the newly introduced field `name` here.

peterbanda commented on web/app/org/ada/web/security/AdaAuthConfig.scala:
```diff
@@ -71,7 +72,7 @@ trait AdaAuthConfig extends AuthConfig {
   override def resolveUser(id: Id)(implicit ctx: ExecutionContext): Future[Option[User]] =
-    userManager.findById(id)
+    userManager.findByOidcId(Option(UUID.fromString(id)))
 
```
Cannot be fixed to OIDC id here (can be undefined)... rollback to `findById`

peterbanda commented on web/test/scala/org/ada/server/controller/AuthControllerTest.scala:
```diff
@@ -0,0 +1,34 @@
+  }
+
+
+
```
empty space

peterbanda reviewed (commented):
The main problem that needs to be addressed is the use of the *optional* `oidcId` attribute as a global id. As the `Option[..]` type suggests, in non-OIDC scenarios this attribute holds `None` value, which will make the code fail at several places. I can see that this change was introduced to address non-uniqueness of `userId` if OIDC-based authentication is used, nevertheless the solution is not to ignore the existing or future non-OIDC authentication scenarios and give up on backward-compatibility.

To sum up the current auth options:
- _Local_:  `userId` - can be pretty much any string (unique), `oidcId` - undefined
- _LDAP_: `userId` - equals to `firstname.lastname` (unique), `oidcId` - undefined
- _OIDC_: `userId` - equals to user name; should be also `firstname.lastname` (potentially non-unique), `oidcId` - "sub" from jwt token (unique)

The first thing to note is that, one instance of Ada won't be using LDAP and OIDC at the same time, therefore we can afford to change the semantics of the fields a bit. More specifically, there must be a single id field that is used within Ada as a unique identifier regardless of the underlying auth provider. I propose this field should be `userId` (of type `String`). For the OIDC auth this requires to feed the "sub" part from jwt token (aka `oidcId`) to `userId`, and dropping the field `oidcId` completely. If we want to preserve the OIDC user name, formerly fed to `userId` (which we should) we can do so by introducing a new optional field - let's say `userName` or `oidcUserName`.

This approach will however have a hidden cost. If the authentication of an existing Ada instance is supposed to be switched from LDAP to OIDC, one would need to migrate the existing users: copy `userId`s to `oidcUserName`s and populate `userId`s from OIDC. This shouldn’t be too hard though.

LBolzani commented:
Good morning Peter,
I've changed the OIDC login and GuiteAppTest classes according to your suggestions.

I remain at your disposal for any further information.

Luca

peterbanda commented:
Hi Luca,

Looks great! It's definitely way to go.

Before I take a closer look pls. make sure you address all the issues I raised one-by-one. I see several of them, e.g. in `PacSecurityModule.scala`, `UserController.scala`, and `navbar.scala.html` (some of them critical) that have been missed out.

Cheers,
:Peter

LBolzani commented:
Good morning Peter,

I did a refactoring on the classes. I didn’t understand what is the problem on _PacSecurityModule.scala_, maybe because is it loaded if LDap configuration has been selected?

Luca

peterbanda commented:
Hi Luca,

Not sure what exactly is not clear but to answer you question: Yes, there could be an LDAP or local auth used without any OIDC configuration param being set.

For reference I am listing all 6 issues related to `PacSecurityModule.scala` here:

1. All the `Option.get` calls are (very) *bad* and should be avoided. If these config settings are not specified by mistake or intentionally because non-OIDC (i.e. LDAP or local) auth is used, the `.get` calls will fail without a reasonable exception.

Instead of that pls. use `.getOrElse(throw new AdaException(s"OIDC config key '${key}' not available."))` (as it was originally done). Actually it's not clear to me why these safeguards have been removed.

2. Instead of the hardcoded url `/oidc/callback` which can be potentially changed in routes we should use the original construction here:

```
val suffix = org.pac4j.play.routes.CallbackController.callback().url
val callbackUrl = baseUrl.replaceAll("/$", "") + suffix
```

3. As mentioned before, the use of OIDC is optional. The original block (as it was implemented before the OIDC v2 was started) with a safe creation of oidcClient based on the presence (or absence) of `oidc.baseUrl` (or as it's now with `oidc.discoveryURI`) should be restored. You don't want to create an OIDC client if there is no OIDC configuration present (i.e., LDAP or local auth is used). Again, I am not sure why it was removed.

4. Revert all the configuration options as they were before, regardless of the current LCSB-specific deployment (with Keycloak). Other facilities that aim to use the OIDC auth will likely have different configs so it's better to be as general as possible here.

The following config options should be reintroduced:

`logoutURL, responseType, scope, clientAuthenticationMethod, preferredJwsAlgorithm` and `useNonce`

5. Because of the never-ending discussion regarding the local-vs-global logout pls. revert the `enableCentralLogout` config (as it was originally done)

6. Remove `callbackController.setDefaultUrl("/?defaulturlafterlogout")`... do you really want to use a String `defaulturlafterlogout`?

Note that these issues are visible in my review and, as mentioned, it's expected that they are fixed (or commented with a reasonable explanation) in an organized fashion.

If you need more help I suggest to contact somebody local (e.g. Sascha).

Thank you for your effort,
:Peter

LBolzani commented:
Good morning Peter,

thank you for the explanation, I've decided to revert _PacSecurityModule_ in your previous implementation.
My refactoring was due to some problems that I've encountered when I was launching the project first times.
Of course was lacking of safer implementation.
Now that I better understanding about the project and OpenId library I reverted in its original version.

I remain at you disposal for any further information.

Luca

peterbanda commented:
Hi Luca,

Thanks for reverting `PacSecurityModule` back to the original (safe) version.
I took a brief look at the rest of the changes and I have only a few comments:

1. A new `User`'s attribute called `userName` is a bit confusing since we already have `userId`, and `name`. Because it's populated only by OIDC it should be renamed to a more appropriate `oidcUserName`.

2. Related to the point 1, in `AppController.scala` a part to obtain a first name:

```
          val userFirstName = user.userName match {
            case Some(userName) => userName.split("\\.", -1).head.capitalize
            case None => user.name
          }
```
should be simply
```
val userFirstName = user.name.split(" ", -1).head.capitalize
```

3. In `UserController` the attribute `userName` (or after renaming `oidcUserName`) can be `None`:

```
       Json.obj("_id" -> user._id, "name" -> user.userName)
```

It should be replaced instead with `name` which is always set (for OIDC as well as for LDAP or local).

4. (optional) In `OidcAuthController` I've noticed that you are (pre)checking whether a user exists by comparing his/her `userId` with a given OIDC `userName`. I understand this is done purely to avoid a full-scale migration (so users are migrated on case-by-case basis) but a clearer solution would be to update all users in a batch and set their new OIDC `userId`s directly in DB by a script (or Ada's runnable).

LBolzani commented:
Good afternoon Peter,

I've applied the code changes. Regarding the last point '4', we're moving definitely with a full scale migration using a batch, like you have suggested, but for the moment we remain temporarily like this.

Luca

peterbanda commented:
Hey Luca,

Looks good, except in the point `2` I wrote:

```
val userFirstName = user.name.split(" ", -1).head.capitalize
```

Since the `name` contains spaces not dots... pls. fix

LBolzani commented:
Good morning Peter,

line code fixed.
 
Thank you.
Luca

peterbanda reviewed (approved):
should be good now
